### DISCUSSION TITLE
I2S support for ESP32

### DISCUSSION THREAD

**Discussion by Stencht** [zephyrproject-rtos/zephyr#52581] [Q&A] [answered]
Hi, 

Another newbie question:  I want to implement I2S into my project using my ESP32-S2 board.
As far as I can tell, there is no support for I2S for the esp32? (i2s interface not defined in the dts, no specific driver for i2s esp32) 

What is the correct zephyr way to implement this? Do I need to:

- adjust the board device tree, add the i2s interface definition?
- create a esp32 driver for i2s ? (under zephyr/drivers/)

Is there a quick & dirty method? (use the i2s driver from espressif? Although I see this depends on freertos)

Thanks

**Comment by cameled** [comment]
Zephyr drivers normally based on MCU's HAL, which provide necessary register wrapper to operate the MCU peripheral.

For a quick fix, you can use that HAL directly. MCU HAL's located at `modules/hal` directory. The HAL usage can refer other esp32 driver, like `i2c_esp32`.

But if you are not such hurry, implement the i2s driver based on [Device Driver Model](https://docs.zephyrproject.org/1.14.1/reference/drivers/index.html#:~:text=The%20Zephyr%20device%20model%20provides%20a%20consistent%20device,I2C%29%20is%20supported%20by%20a%20generic%20type%20API) is a good choice. With standard driver, you can easily to get support from zephyr community, easy to maintaince, and so on.

**Comment by rftafas** [comment, marked as answer]
> As far as I can tell, there is no support for I2S for the esp32? (i2s interface not defined in the dts, no specific driver for i2s esp32)

Indeed, no support yet. A list of supported peripheral is [here](https://github.com/zephyrproject-rtos/zephyr/issues/29394).

> What is the correct zephyr way to implement this? Do I need to:
> 
>     adjust the board device tree, add the i2s interface definition?
>     create a esp32 driver for i2s ? (under zephyr/drivers/)

Yes, you would have to do both.

> Is there a quick & dirty method? (use the i2s driver from espressif? Although I see this depends on freertos)

I would not recommend that. You may use the hal directly, but have in mind that this is not a public API and it may change without notice.
If you are in a hurry and it is for professional application, I can only suggest:

- Get hands on and write the driver. In this case, we are favoring using hal_espressif instead direct register access.
- Move to ESP-IDF, knowing we currently do not support a migration back to Zephyr without re-flashing the whole device (i.e. using cable, no OTA).

**Comment by Stencht** [comment]
Thanks for the replies, my apologies for the very late response. 

FYI: Eventually I moved to ESP-IDF for the project.
